RIOT-OS/RIOT
# Implement support for link time reordering

### Contribution description

Link time reordering is a protection mechanisms, implemented at compile-time,
which randomizes the memory layout of the resulting firmware image. As such, it
serves as a protection mechanisms against remote code executions which attempt
to reuse existing code. Consider, for example, the PoC exploit in #10739. This
exploit jumps to a GPIO pin toggle function in the text segment. If text
segment addresses differ for each build this would be more difficult to achieve
in a portable way.

Link time reordering is further described in a publication by Koviu et al.
[\[1, p. 394\]][1]. The technique is also utilized by some conventional operating
systems. For instance, OpenBSD uses this technique for randomizing text segment
addresses of the kernel, they refer to it as kernel address randomized link (KARL) [\[2\]][2].

### Testing procedure

Verify that the text segement addresses differ for each build:

```
$ make -C examples/gnrc_networking
$ nm examples/gnrc_networking/bin/native/gnrc_networking.elf > /tmp/first_build
$ make -C examples/gnrc_networking
$ nm examples/gnrc_networking/bin/native/gnrc_networking.elf > /tmp/second_build
$ diff -u /tmp/first_build /tmp/second_build
```

Also make sure they don't when building with `RIOTINSECURE=1`.

### Issues/PRs references

* #13175 

[1]: https://ieeexplore.ieee.org/abstract/document/7917121/
[2]: https://marc.info/?l=openbsd-tech&m=149732026405941
